File: audira_frontend/lib/core/models/artist_metrics_summary.dart
import 'package:equatable/equatable.dart';

/// Summary metrics for an artist
/// GA01-108: Resumen rápido (plays, valoraciones, ventas, comentarios, evolución)
class ArtistMetricsSummary extends Equatable {
  final int artistId;
  final String artistName;
  final DateTime generatedAt;

  // Plays metrics
  final int totalPlays;
  final int playsLast30Days;
  final double playsGrowthPercentage;

  // Rating metrics
  final double averageRating;
  final int totalRatings;
  final double ratingsGrowthPercentage;

  // Sales metrics
  final int totalSales;
  final double totalRevenue;
  final int salesLast30Days;
  final double revenueLast30Days;
  final double salesGrowthPercentage;
  final double revenueGrowthPercentage;

  // Comments metrics
  final int totalComments;
  final int commentsLast30Days;
  final double commentsGrowthPercentage;

  // Content metrics
  final int totalSongs;
  final int totalAlbums;
  final int totalCollaborations;

  // Top performing
  final int? mostPlayedSongId;
  final String? mostPlayedSongName;
  final int? mostPlayedSongPlays;

  const ArtistMetricsSummary({
    required this.artistId,
    required this.artistName,
    required this.generatedAt,
    required this.totalPlays,
    required this.playsLast30Days,
    required this.playsGrowthPercentage,
    required this.averageRating,
    required this.totalRatings,
    required this.ratingsGrowthPercentage,
    required this.totalSales,
    required this.totalRevenue,
    required this.salesLast30Days,
    required this.revenueLast30Days,
    required this.salesGrowthPercentage,
    required this.revenueGrowthPercentage,
    required this.totalComments,
    required this.commentsLast30Days,
    required this.commentsGrowthPercentage,
    required this.totalSongs,
    required this.totalAlbums,
    required this.totalCollaborations,
    this.mostPlayedSongId,
    this.mostPlayedSongName,
    this.mostPlayedSongPlays,
  });

  factory ArtistMetricsSummary.fromJson(Map<String, dynamic> json) {
    return ArtistMetricsSummary(
      artistId: (json['artistId'] as num).toInt(),
      artistName: json['artistName'] as String,
      generatedAt: DateTime.parse(json['generatedAt'] as String),
      totalPlays: (json['totalPlays'] as num).toInt(),
      playsLast30Days: (json['playsLast30Days'] as num).toInt(),
      playsGrowthPercentage: (json['playsGrowthPercentage'] as num).toDouble(),
      averageRating: (json['averageRating'] as num).toDouble(),
      totalRatings: (json['totalRatings'] as num).toInt(),
      ratingsGrowthPercentage:
          (json['ratingsGrowthPercentage'] as num).toDouble(),
      totalSales: (json['totalSales'] as num).toInt(),
      totalRevenue: (json['totalRevenue'] as num).toDouble(),
      salesLast30Days: (json['salesLast30Days'] as num).toInt(),
      revenueLast30Days: (json['revenueLast30Days'] as num).toDouble(),
      salesGrowthPercentage: (json['salesGrowthPercentage'] as num).toDouble(),
      revenueGrowthPercentage:
          (json['revenueGrowthPercentage'] as num).toDouble(),
      totalComments: (json['totalComments'] as num).toInt(),
      commentsLast30Days: (json['commentsLast30Days'] as num).toInt(),
      commentsGrowthPercentage:
          (json['commentsGrowthPercentage'] as num).toDouble(),
      totalSongs: (json['totalSongs'] as num).toInt(),
      totalAlbums: (json['totalAlbums'] as num).toInt(),
      totalCollaborations: (json['totalCollaborations'] as num).toInt(),
      mostPlayedSongId: json['mostPlayedSongId'] != null ? (json['mostPlayedSongId'] as num).toInt() : null,
      mostPlayedSongName: json['mostPlayedSongName'] as String?,
      mostPlayedSongPlays: json['mostPlayedSongPlays'] != null ? (json['mostPlayedSongPlays'] as num).toInt() : null,
    );
  }

  @override
  List<Object?> get props => [
        artistId,
        artistName,
        generatedAt,
        totalPlays,
        playsLast30Days,
        playsGrowthPercentage,
        averageRating,
        totalRatings,
        ratingsGrowthPercentage,
        totalSales,
        totalRevenue,
        salesLast30Days,
        revenueLast30Days,
        salesGrowthPercentage,
        revenueGrowthPercentage,
        totalComments,
        commentsLast30Days,
        commentsGrowthPercentage,
        totalSongs,
        totalAlbums,
        totalCollaborations,
        mostPlayedSongId,
        mostPlayedSongName,
        mostPlayedSongPlays,
      ];
}

File: music-catalog-service/src/main/java/io/audira/catalog/dto/ArtistMetricsSummary.java
package io.audira.catalog.dto;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * DTO que encapsula el resumen ejecutivo de métricas para un artista.
 * <p>
 * Diseñado para alimentar el "Dashboard" principal del artista, cumpliendo con el requisito
 * <b>GA01-108: Resumen rápido</b>. Proporciona una "foto" del estado actual, incluyendo
 * acumulados históricos, rendimiento reciente (últimos 30 días) y porcentajes de crecimiento
 * comparativos (vs. periodo anterior).
 * </p>
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ArtistMetricsSummary {

    // --- Información Básica ---
    /** Identificador único del artista al que pertenecen las métricas. */
    private Long artistId;

    /** Nombre artístico (desnormalizado para evitar consultas extra). */
    private String artistName;

    /**
     * Fecha y hora exacta en la que se calculó este resumen.
     * <p>Útil para mostrar al usuario la frescura de los datos (ej: "Actualizado hace 5 min").</p>
     */
    private LocalDateTime generatedAt;

    // --- Métricas de Reproducción (Plays) ---
    /** Número total histórico de reproducciones de todas las canciones. */
    private Long totalPlays;

    /** Reproducciones acumuladas exclusivamente en los últimos 30 días. */
    private Long playsLast30Days;

    /**
     * Porcentaje de crecimiento de reproducciones.
     * <p>Comparación: (Últimos 30 días) vs (Periodo de 30 días anterior).</p>
     * <p>Puede ser negativo si el rendimiento ha bajado.</p>
     */
    private Double playsGrowthPercentage;

    // --- Métricas de Valoración (Ratings) ---
    /** Promedio de estrellas (1.0 a 5.0) de todo el catálogo del artista. */
    private Double averageRating;

    /** Cantidad total de votos recibidos. */
    private Long totalRatings;

    /** Variación porcentual de la valoración promedio respecto al mes anterior. */
    private Double ratingsGrowthPercentage;

    // --- Métricas de Ventas y Ganancias ---
    /** Número total de unidades vendidas (Canciones + Álbumes). */
    private Long totalSales;

    /** Ingresos totales históricos generados (antes de splits). */
    private BigDecimal totalRevenue;

    /** Unidades vendidas en los últimos 30 días. */
    private Long salesLast30Days;

    /** Ingresos generados en los últimos 30 días. */
    private BigDecimal revenueLast30Days;

    /** Porcentaje de crecimiento en volumen de ventas. */
    private Double salesGrowthPercentage;

    /** Porcentaje de crecimiento en ingresos monetarios. */
    private Double revenueGrowthPercentage;

    // --- Métricas de Interacción (Comentarios) ---
    /** Total histórico de comentarios recibidos. */
    private Long totalComments;

    /** Comentarios recibidos en el último mes. */
    private Long commentsLast30Days;

    /** Variación en la actividad de comentarios. */
    private Double commentsGrowthPercentage;

    // --- Métricas de Inventario de Contenido ---
    /** Número de canciones activas en el catálogo. */
    private Long totalSongs;

    /** Número de álbumes publicados. */
    private Long totalAlbums;

    /** Número de obras donde figura como colaborador invitado. */
    private Long totalCollaborations;

    // --- Contenido Destacado (Top Performing) ---
    /** ID de la canción con mayor número de reproducciones históricas. */
    private Long mostPlayedSongId;
    
    /** Título de la canción más escuchada. */
    private String mostPlayedSongName;
    
    /** Número total de reproducciones de la canción más escuchada. */
    private Long mostPlayedSongPlays;
}
